L1Trigger/L1CaloTrigger/interface/Phase2L1CaloPFClusterEmulator.h

```diff
@@ -0,0 +1,197 @@
+// 5             |                                                     |
+//               | ----------------------------------------------------|
+// 6             |                                                     |
+// 7             |                                                     |
+//
+// 8 PFclusters are created in one 21x8 (2+17+2 x 2+4+2)
+
+static constexpr int nTowerEta = 34;
+static constexpr int nTowerPhi = 72;
+static constexpr int nSLR = 36;
+static constexpr int nTowerEtaSLR = 21;  // including overlap: 2+17+2
+static constexpr int nTowerPhiSLR = 8;   // including overlap: 2+4+2
+static constexpr int nPFClusterSLR = 8;
+
+namespace gctpf {
+
+  typedef struct {
+    float et;
+    int eta;
+    int phi;
+  } GCTpfcluster_t;
+
+  typedef struct {
+    GCTpfcluster_t GCTpfclusters[nPFClusterSLR];
+  } GCTPfcluster_t;
+
+  typedef struct {
+    float et;
+    int eta;
+    int phi;
+  } GCTint_t;
+
+  typedef struct {
+    GCTint_t t[nTowerPhiSLR];
+  } gctEtaStrip_t;
+
+  typedef struct {
+    GCTint_t p[nTowerEtaSLR - 2];
+  } gctEtaStripPeak_t;
+
+  typedef struct {
+    gctEtaStrip_t s[nTowerEtaSLR];
+  } region_t;
+
+  inline GCTint_t bestOf2(const GCTint_t& t0, const GCTint_t& t1) {
+    GCTint_t x;
+    x = (t0.et > t1.et) ? t0 : t1;
+
+    return x;
+  }
+
+  inline GCTint_t getPeakOfStrip(const gctEtaStrip_t& etaStrip) {
+    GCTint_t best12 = bestOf2(etaStrip.t[1], etaStrip.t[2]);
+    GCTint_t best34 = bestOf2(etaStrip.t[3], etaStrip.t[4]);
+    GCTint_t best56 = bestOf2(etaStrip.t[5], etaStrip.t[6]);
+    GCTint_t best1234 = bestOf2(best12, best34);
+    GCTint_t bestAll = bestOf2(best1234, best56);
+
+    return bestAll;
+  }
+
+  inline GCTint_t getPeakBin(const gctEtaStripPeak_t& etaStripPeak) {
+    GCTint_t best01 = bestOf2(etaStripPeak.p[0], etaStripPeak.p[1]);
+    GCTint_t best23 = bestOf2(etaStripPeak.p[2], etaStripPeak.p[3]);
+    GCTint_t best45 = bestOf2(etaStripPeak.p[4], etaStripPeak.p[5]);
+    GCTint_t best67 = bestOf2(etaStripPeak.p[6], etaStripPeak.p[7]);
+    GCTint_t best89 = bestOf2(etaStripPeak.p[8], etaStripPeak.p[9]);
+    GCTint_t best1011 = bestOf2(etaStripPeak.p[10], etaStripPeak.p[11]);
+    GCTint_t best1213 = bestOf2(etaStripPeak.p[12], etaStripPeak.p[13]);
+    GCTint_t best1415 = bestOf2(etaStripPeak.p[14], etaStripPeak.p[15]);
+    GCTint_t best1617 = bestOf2(etaStripPeak.p[16], etaStripPeak.p[17]);
+    GCTint_t best0123 = bestOf2(best01, best23);
+    GCTint_t best4567 = bestOf2(best45, best67);
+    GCTint_t best891011 = bestOf2(best89, best1011);
+    GCTint_t best12131415 = bestOf2(best1213, best1415);
+    GCTint_t best01234567 = bestOf2(best0123, best4567);
+    GCTint_t best01234567891011 = bestOf2(best01234567, best891011);
+    GCTint_t best121314151617 = bestOf2(best12131415, best1617);
+    GCTint_t best12131415161718 = bestOf2(best121314151617, etaStripPeak.p[18]);
+    GCTint_t bestAll = bestOf2(best01234567891011, best12131415161718);
+
```

Comment:
Hi Andrea, since this code is firmware based, it will be easier to keep the current structure for following future changes in firmware.

Reply:
> Hi Andrea, since this code is firmware based, it will be easier to keep the current structure for following future changes in firmware.

It is ugly...
But ok, if the firmware is made as such let stick on it.
Let hope that not so many updates will be needed in the future